io/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -370,9 +360,8 @@ sealed abstract class IO[E, A] { self =>
    * Retries continuously until the action succeeds or the specified duration
    * elapses.
    */
-  final def retryFor(duration: Duration): IO[E, Maybe[A]] =
-    retry.map(Maybe.just[A]) race
-      (IO.sleep[E](duration) *> IO.now[E, Maybe[A]](Maybe.empty[A]))
+  final def retryFor[B](z: B)(f: A => B)(duration: Duration): IO[E, B] =
```

Comment:
This is an improvement, more polymorphism. 😄

Reply:
@jdegoes 

1. Should we put `z` and `f` in the same args list to avoid type inference nuisances?
2. Should we provide a shortcut for an `Option` version equivalent to the previous API?

Reply:
1. Yes, if it actually improves inference. 
2. I am not sure it's necessary since it's so easy to do `retryFor(None)(Some)`.